packages/dd-trace/src/appsec/rule_manager.js

```diff
@@ -3,16 +3,44 @@
 const callbacks = require('./callbacks')
 const Gateway = require('./gateway/engine')
 
-const appliedCallbacks = new Map()
+let appliedCallback
 const appliedAsmData = new Map()
+let defaultRules
+let asmDDRules
+let config
 
-function applyRules (rules, config) {
-  if (appliedCallbacks.has(rules)) return
+function applyRules (rules, _config) {
+  if (appliedCallback) return
+  defaultRules = rules
+  config = _config
+  appliedCallback = initDDWAF(rules)
+}
 
-  // for now there is only WAF
-  const callback = new callbacks.DDWAF(rules, config)
+function initDDWAF (rules) {
+  return new callbacks.DDWAF(rules, config)
+}
 
-  appliedCallbacks.set(rules, callback)
+function updateAsmDDRules (action, asmRules) {
+  if (action === 'unapply') {
+    asmDDRules = undefined
+  } else {
+    asmDDRules = asmRules
+  }
+  updateAppliedRules()
+  updateAppliedRuleData()
+}
+
+function updateAppliedRules () {
+  const rules = asmDDRules || defaultRules
+  try {
+    const deprecatedCallback = appliedCallback
+    appliedCallback = initDDWAF(rules)
```

Comment:
I think you might need to do a Gateway.manager.clear() here too, otherwise the old WAF callback will still be called no ?

Reply:
I did it in line 50. Out of the try catch block.